web-animations/timing-model/timelines/document-timelines.html

```diff
@@ -9,6 +9,10 @@
 <script>
 'use strict';
 
+function matchUnconditionalClamping(timestamp) {
+  return parseFloat((Math.floor(timestamp / .02) * .02).toPrecision(8), 10);
```

Comment:
@tomrittervg is the clamping spec'ed? `document.timeline.currentTime` (behind a flag) has a precision of 1e-3 in Chrome so this change breaks the test in Chrome cc @flackr .

Reply:
As far as I know the only statement about time precision is the web animations spec [recommending user agents be able to represent time values with microsecond precision](https://drafts.csswg.org/web-animations/#precision-of-time-values) (i.e. 3 decimal places on milliseconds). We have used this (1e-3) as a tolerance for precision errors in other wpt tests: https://github.com/web-platform-tests/wpt/search?q=assert_times_equal&unscoped_q=assert_times_equal

I'm a bit confused by why document timeline should be less precise than requestAnimationFrame?

Reply:
So this is my first experience with web-platform tests and I was advised to just edit the tests in our tree... 

In the wake of Spectre, most browsers clamped (and jittered) performance.now() with a resolution of 1ms. However the mitigations were only applied to performance.now() and not all other timers (including ones that exposed highly precise timestamps.) Firefox applied those mitigations to **all** timers except Animation-based ones for annoying reasons.

We finally got around to doing some clamping on Animation-based timers, using a value of 20 microseconds.

This behavior doesn't match other browsers. I was advised to guard some checks (in timing-override.js ) with a firefox guard; if I put a similar guard on this function is that the best path forward?

Reply:
I'm not sure what you mean by a firefox guard but I would advise against this for a WPT test.

I think the right thing to do is to expect document timeline and requestAnimationFrame timestamp to either round to the same value when rounded to something less precise than 20 microseconds, or to expect the values to be within some allowed delta with a comment that while the recommendation is for time values to be precise enough to have microsecond precision it is not required and some browsers choose to use less precision to mitigate the risk of timing attacks.

Reply:
On a sidenote, why do requestAnimationFrame timestamps in Firefox not have the same clamping? The spec notes that [DocumentTimeline only updates when we update animations and send events](https://drafts.csswg.org/web-animations/#the-documents-default-timeline) so it doesn't need to update more frequently than requestAnimationFrames fire.

Reply:
FWIW I approved the patch, which as I understood it allows Firefox (only) to pass the test with a lower timer resoultion than other browsers. In general I agree that having browser-specific code in these tests is bad, but in some cases it seems like a reasonable tradeoff (e.g. to the extent that it's web-compatible at all it doesn't make sense to fail animations tests just because lower timer precision affects the acceptable delta when comapring results).

Reply:
To be clear, I'm not at all opposed to reducing the expected timer resolution so that Firefox will still pass the test, the issue with this change is that only the requestAnimationFrame timestamp is clamped. This means in browsers which don't have the reduced precision it fails as the actual value is more precise. For example, in a recent test run:

FAIL Document timelines report current time relative to navigationStart assert_equals: The current time matches requestAnimationFrame time expected 78.26 but got 78.277

I think the right thing is to either clamp both the actual and expected values, or round both, or allow some delta between them as we do for time comparisons with assert_times_equal

Reply:
See https://github.com/web-platform-tests/wpt/pull/18750